### PR TITLE
KAA-966 Add ability to set failover strategies in Objective-C EP SDK

### PR DESCRIPTION
**Pull request by abohomol**

### CODE REVIEW

File: client/client-multi/client-objective-c/Kaa/channel/failover/DefaultFailoverManager.m [hunk 1]
```diff
@@ -16,15 +16,14 @@
 
 #import "DefaultFailoverManager.h"
 #import "KaaLogging.h"
+#import "DefaultFailoverStrategy.h"
+#import "KaaExceptions.h"
 
-#define TAG @"DefaultFailoverManager >>>"
+static const NSString *logTag = @"DefaultFailoverManager >>>";
```
Comment (sergiyk0) [inline client/client-multi/client-objective-c/Kaa/channel/failover/DefaultFailoverManager.m]: const here doesn't do anything really, NSString is immutable by itself. Maybe you wanted to say "static NSString *const"  as in constant pointer?

File: client/client-multi/client-objective-c/Kaa/channel/failover/DefaultFailoverManager.m [hunk 2]
```diff
@@ -166,129 +157,97 @@ - (void)onServerChangedWithConnectionInfo:(id<TransportConnectionInfo>)connectio
...
 }
 
 - (FailoverDecision *)decisionOnFailoverStatus:(FailoverStatus)status {
     
-    DDLogInfo(@"%@ Applying failover strategy for status: %i", TAG, status);
+    DDLogVerbose(@"%@ Applying failover decision for status: %i", logTag, status);
 
     @synchronized(self) {
-        NSNumber *serverTypeKey = nil;
+        AccessPointIdResolution *accessPointIdResolution = nil;
```
Comment (sergiyk0) [inline client/client-multi/client-objective-c/Kaa/channel/failover/DefaultFailoverManager.m:205]: Can't see where this is variable is used. It is created and modified, but is it passed anywhere?

File: client/client-multi/client-objective-c/Kaa/channel/failover/strategies/DefaultFailoverStrategy.m [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#import "DefaultFailoverStrategy.h"
+#import "KaaLogging.h"
+
+static const NSString *logTag = @"DefaultFailoverStrategy >>>";
```
Comment (sergiyk0) [inline client/client-multi/client-objective-c/Kaa/channel/failover/strategies/DefaultFailoverStrategy.m]: Probably you wanted to create constant pointer here as well.

File: client/client-multi/client-objective-c/Kaa/channel/failover/DefaultFailoverManager.m (hunk 2 above)
Comment (abohomol) [inline client/client-multi/client-objective-c/Kaa/channel/failover/DefaultFailoverManager.m:205]: It's modified version is used in other DefaultFailoverManager methods.

File: client/client-multi/client-objective-c/Kaa/channel/failover/DefaultFailoverManager.m (hunk 1 above)
Comment (abohomol) [inline client/client-multi/client-objective-c/Kaa/channel/failover/DefaultFailoverManager.m]: Yes, exactly: https://github.com/kaaproject/kaa/pull/582/commits/2f4c3b879e47572c1cfc78f628913d6da649061b

File: client/client-multi/client-objective-c/Kaa/channel/failover/DefaultFailoverManager.m (hunk 2 above)
Comment (sergiyk0) [inline client/client-multi/client-objective-c/Kaa/channel/failover/DefaultFailoverManager.m:205]: OK, now I get it.

File: client/client-multi/client-objective-c/Kaa/channel/failover/DefaultFailoverManager.m [hunk 4]
```diff
@@ -99,30 +90,30 @@ - (instancetype)initWithChannelManager:(id<KaaChannelManager>)channelMgr
...
     }
     
     @synchronized(self) {
         long currentResolutionTime = -1;
         AccessPointIdResolution *pointResolution = self.resolutionProgressMap[@([connectionInfo serverType])];
         if (pointResolution) {
-            currentResolutionTime = pointResolution.resolutionTime;
+            currentResolutionTime = pointResolution.resolutionTimeMillis;
             if (pointResolution.accessPointId == [connectionInfo accessPointId]
                 && pointResolution.resolution
```
Comment (sergiyk0) [inline client/client-multi/client-objective-c/Kaa/channel/failover/DefaultFailoverManager.m]: Please use explicit comparison with nil when checking if object exists.

File: client/client-multi/client-objective-c/Kaa/channel/failover/strategies/DefaultFailoverStrategy.m [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+- (void)onRecoverWithConnectionInfo:(id<TransportConnectionInfo>)connectionInfo {
+    DDLogDebug(@"%@ SDK recovered after failover with connection info: %@", logTag, connectionInfo);
+}
+
+- (FailoverDecision *)decisionOnFailoverStatus:(FailoverStatus)status {
+    DDLogVerbose(@"%@ Producing failover decision for failover status: %i", logTag, status);
+    switch (status) {
+        case FAILOVER_STATUS_BOOTSTRAP_SERVERS_NA:
+            return [[FailoverDecision alloc] initWithFailoverAction:FAILOVER_ACTION_RETRY
```
Comment (sergiyk0) [inline client/client-multi/client-objective-c/Kaa/channel/failover/strategies/DefaultFailoverStrategy.m:68]: retryPeriod is passed without converting to milliseconds

Comment (abohomol) [inline client/client-multi/client-objective-c/Kaa/channel/failover/strategies/DefaultFailoverStrategy.m:68]: +1 for attention - [fixed](https://github.com/kaaproject/kaa/pull/582/commits/bc27421f6df4a9f96e2daf65b9706d9b531a3f3f).

File: client/client-multi/client-objective-c/Kaa/channel/failover/strategies/DefaultFailoverStrategy.m [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+                                                        retryPeriod:_bootstrapServersRetryPeriod
```
Comment (sergiyk0) [inline client/client-multi/client-objective-c/Kaa/channel/failover/strategies/DefaultFailoverStrategy.m]: Why ivars here are accessed with underscore? It is undesirable for code style

Comment (abohomol) [inline client/client-multi/client-objective-c/Kaa/channel/failover/strategies/DefaultFailoverStrategy.m]: Fixed.

**Comment by sergiyk0** [comment]
Looks good to me

**Pull request merged**